sdk/storage/Azure.Storage.Blobs/src/BlobDirectoryClient.cs

```diff
@@ -0,0 +1,726 @@
+        /// </param>
+        /// <param name="options">
+        /// Optional Parameters <see cref="BlobUploadDirectoryOptions"/>
+        /// </param>
+        /// <param name="async">
+        /// Whether to invoke the operation asynchronously.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response{BlobContentInfo}"/> describing the
+        /// state of the updated block blob.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        /// TODO: remove pragma warning after adding await operators
+#pragma warning disable CS1998 // Async method lacks 'await' operators and will run synchronously
+        internal virtual async Task<Response<BlobContentInfo>> UploadInternal(
+#pragma warning restore CS1998 // Async method lacks 'await' operators and will run synchronously
+            string directory,
+            BlobUploadDirectoryOptions options,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            //TODO: Implement directory upload
+            ResponseWithHeaders<BlockBlobUploadHeaders> response = ResponseWithHeaders.FromValue((BlockBlobUploadHeaders)default, default);
+            return Response.FromValue(response.ToBlobContentInfo(), response.GetRawResponse());
+        }
+        #endregion
+
+        #region Download
+        /// <summary>
+        /// The <see cref="Download(string)"/> operation downloads a blob using parallel requests,
+        /// and writes the content to <paramref name="path"/>.
+        /// </summary>
+        /// <param name="path">
+        /// A file path to write the downloaded content to.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response"/> describing the operation.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public virtual Response Download(string path) =>
+            Download(path, CancellationToken.None);
+
+        /// <summary>
+        /// The <see cref="DownloadAsync(string)"/> downloads a blob using parallel requests,
+        /// and writes the content to <paramref name="path"/>.
+        /// </summary>
+        /// <param name="path">
+        /// A file path to write the downloaded content to.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response"/> describing the operation.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public virtual async Task<Response> DownloadAsync(string path) =>
```

Comment:
should this be `targetPath`? i.e. is this path refereing to local file system or virtual path in storage account?

similarly the response type probably doesn't make sense like for upload api. we should offer callback/events.